### ISSUE TITLE
scalaInstance is created for module even if autoScalaLibrary is set to false

### ISSUE THREAD

**Issue by unkarjedy** [sbt/sbt#6559]
## steps

1. create sbt project:
```scala
name := "untitled57"

version := "0.1"

scalaVersion := "2.13.6"

lazy val module1 = project in file("module1")
lazy val module2 = (project in file("module2")).settings(autoScalaLibrary := false)
```

2. create scala file in each project:
![image](https://user-images.githubusercontent.com/3989292/122889265-4f37d880-d34b-11eb-8846-c5e19f3e9ccd.png)

3. open sbt 
4. run `show scalaInstance`

## problem
See that scalaInstance is reported for `module2` as well:
![image](https://user-images.githubusercontent.com/3989292/122889726-b5bcf680-d34b-11eb-848f-869f3c075a69.png)

even though `compile` would fail for the module:
![image](https://user-images.githubusercontent.com/3989292/122889903-e3a23b00-d34b-11eb-8118-e0ac50edfcc3.png)

## expectation
Looks like `scalaIntance` shouldn't be created for `module2` in this case

**Comment by eed3si9n** [comment]
I'm not sure if I agree with the expectation here. `autoScalaLibrary` means don't add scala-library automatically, which means that the build user can add them explicitly. It is often used as an indicator to mean that a subproject is not a Scala subproject, you can in theory construct your own version of scala-library etc, but use some version of Scala compiler, so `scalaInstance` and `autoScalaLibrary` are independent, I think.

Given that `scalaInstance` is a task, it will not be executed until you ask for it, and by typing in `show scalaInstance`, you just asked for it.

**Comment by eed3si9n** [comment]
Also please see https://www.scala-sbt.org/1.x/docs/Basic-Def.html#How+build.sbt+defines+settings

`scalaVersion` should be scoped to `ThisBuild / scalaVersion` if you intend it to be used by module1 and module2.

**Comment by unkarjedy** [comment]
>scalaVersion should be scoped to ThisBuild / scalaVersion if you intend it to be used by module1 and module2.

Yes, thanks, I know that it fallbacks  to the Scala version of sbt itself to I had no questions regarding `module1`

> Given that scalaInstance is a task, it will not be executed until you ask for it, and by typing in show scalaInstance, you just asked for it

In IntelliJ Scala Plugin (via https://github.com/JetBrains/sbt-structure)
we use `show scalaInstance` to understand whether the given module "is Scala module" and whether we should configure Scala SDK for the module

>build user can add them explicitly

Maybe do not construct `scalaIntance` if there is no appropriate scala library explicitly added by the user?
`scalaInstance` returns a perfectly valid instance for `module2`. 
This is misleading because `sbt compile` fails, telling that it can't find the `scala.native` staff, while it's there in `scalaInstance`

**Comment by eed3si9n** [comment]
> we use `show scalaInstance` to understand whether the given module "is Scala module" and whether we should configure Scala SDK for the module

As far as I know, there's no setting right now to say if a JvmPlugin subproject is a Scala module or not. This makes sense I think because even if you only had Java sources, we would still need to analyze them for incremental compilation and pipelining purposes.

**Comment by unkarjedy** [comment]
>we would still need to analyze them for incremental compilation 

Wouldn't/shouldn't that be done by the `scalaInstance` of the module which depends on this java module?

**Comment by eed3si9n** [comment]
Scala compiler has the ability to parse Java sources and also generate type information (pickle) out of them. Also Zinc phases are injected in Scala compiler, but not Java compiler. Things could probably work even if we don't do these things (like treating Java as external JAR), but it could result to loss of functionality.

**Comment by unkarjedy** [comment]
>Scala compiler has the ability to parse Java sources and also generate type information (pickle) out of them

Can this logic in theory be different between scala versions?
E.g. there is some bug in one scala version and not in another?
Or maybe the pickle format is different between scala versions?
